### PR TITLE
boost-build: 2016.03 -> 4.4.1

### PR DESCRIPTION
**Pull request by dotlambda**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @itkatchev

### CODE REVIEW

File: pkgs/development/tools/boost-build/default.nix [hunk 1]
```diff
@@ -11,8 +11,6 @@ stdenv.mkDerivation rec {
     sha256 = "1qw5marmp7z09nwcjlqrmqdg9b6myfqj3zvfz888x9mbidrmhn6p";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/boost-build/default.nix:11]: I think this needs an update.

Comment (dotlambda) [reply pkgs/development/tools/boost-build/default.nix:11]: oops

**Comment by dotlambda** [comment]
I have no idea how to make it install bjam.

**Comment by dotlambda** [comment]
Turns out bjam was turned into an alias and - I guess - removed later: https://boostorg.github.io/build/manual/master/index.html#bbv2.faq.names

**Comment by dotlambda** [comment]
Since some other packages still use bjam, I'll add a symlink.

**Review by oxalica — APPROVED** [review]
6 packages built on x86_64-linux:
boost-build luabind luabind_luajit osmscout-server osrm-backend soi

Not sure if the version "downgrade" will cause issues in nix-env.

File: pkgs/development/tools/boost-build/default.nix [hunk 2]
```diff
@@ -1,34 +1,39 @@
...
+    runHook preBuild
     ./bootstrap.sh
+    runHook postBuild
   '';
 
   installPhase = ''
-    ./b2 install --prefix=$out
+    runHook preInstall
+    ./b2 install --prefix="$out"
+    ln -s b2 "$out/bin/bjam"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/boost-build/default.nix:31]:
```suggestion
    ln -s $out/bin/b2 $out/bin/bjam
```

Comment (dotlambda) [reply pkgs/development/tools/boost-build/default.nix:31]: That really doesn't have any advantage.

Comment (oxalica) [reply pkgs/development/tools/boost-build/default.nix:31]: I don't think we need absolute path here. Symbolic link target is always related to itself. 
We also have `{cp,mv,ln} -> coreutils` in `coreutils`.

Comment (dotlambda) [reply pkgs/development/tools/boost-build/default.nix:31]: Turns out the `fixupPhase` prefers relative symlinks anyway:
```
rewriting symlink /nix/store/z1cx1qsjqyx9dzxsxd67cn1jm3mwhc71-boost-build-4.4.1/bin/bjam to be relative to /nix/store/z1cx1qsjqyx9dzxsxd67cn1jm3mwhc71-boost-build-4.4.1
```

**Pull request merged**